comit/src/expiries.rs

```diff
@@ -0,0 +1,631 @@
+            }
+            return AliceAction::Abort;
+        }
+
+        let (next_action, _state) = current_state.next();
+        next_action
+    }
+
+    /// Returns the recommended next action that Bob should take.
+    pub fn next_action_for_bob(&self, current_state: BobState) -> BobAction {
+        if current_state == BobState::Done {
+            return BobAction::NoFurtherAction;
+        }
+
+        let funded = current_state.has_funded_beta();
+
+        if self.beta_expiry_has_elapsed() {
+            if funded {
+                return BobAction::Refund;
+            }
+            return BobAction::Abort;
+        };
+
+        if !self.bob_has_time_to_complete(current_state) {
+            if funded {
+                return BobAction::WaitForRefund;
+            }
+            return BobAction::Abort;
+        }
+
+        let (next_action, _state) = current_state.next();
+        next_action
+    }
+
+    fn alpha_expiry_has_elapsed(&self) -> bool {
+        let now = self.alpha_connector.current_time();
+        now > self.alpha_expiry
+    }
+
+    fn beta_expiry_has_elapsed(&self) -> bool {
+        let now = self.beta_connector.current_time();
+        now > self.beta_expiry
+    }
+}
+
+/// The minimum time we should allow for Alice to transition from
+/// `current_state` to done.
+fn period_for_alice_to_complete(config: &Config, current_state: AliceState) -> Duration {
+    // Tail call recursion with an accumulator.
+    fn period_to_complete(config: &Config, state: AliceState, acc: Duration) -> Duration {
+        if state == AliceState::Done {
+            return acc;
+        }
+
+        let (_action, next_state) = state.next();
+        let transition_period = state.transition_period(config);
+
+        period_to_complete(config, next_state, acc + transition_period)
+    }
+
+    period_to_complete(config, current_state, Duration::zero())
+}
+
+/// The minimum time we should allow for Bob to transition from
+/// `current_state` to done.
+fn period_for_bob_to_complete(config: &Config, current_state: BobState) -> Duration {
+    // Tail call recursion with an accumulator.
+    fn period_to_complete(config: &Config, state: BobState, acc: Duration) -> Duration {
+        if state == BobState::Done {
+            return acc;
+        }
+
+        let (_action, next_state) = state.next();
+        let transition_period = state.transition_period(config);
+
+        period_to_complete(config, next_state, acc + transition_period)
+    }
+
+    period_to_complete(config, current_state, Duration::zero())
+}
+
+/// Happy path states for Alice.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum AliceState {
+    // FIXME: Do we need a pre-created state in order to give Alice one more TTA period
+    // before `created`?
+    /// Initial state, swap has been created.
+    Created,
+    /// The fund alpha transaction has been sent to the network.
+    FundAlphaTransactionSent,
```

Comment:
Can we confirm we only assume a local state to reach this variant? e.g. Bob never considers Alice state to be this variant?

Reply:
Not sure what you are asking. Code now includes the following rustdoc
```
/// From<'role'State> converts a state object into a best guess at the
/// counterparties state. We cannot know for sure what state the counterparty is
/// in because there are state transitions which rely on local knowledge. We
/// make a conservative guess at the counterparties state using only public
/// knowledge.
```

Reply:
So are you saying we are going to look into the Bitcoin and Ethereum mempool to be able to transition the counterpart state to `FundAlphaTransactionSent`?